### PR TITLE
MinGW: Enable C++ threads with mcfgthreads

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

This uses https://github.com/lhmouse/mcfgthread to get C++ threads with mingw without the slow POSIX threads emulation. The bootstrapping infrastructure is set up so this is factored through `threadsCross` attribute which could be parameterized to support both gthreads (GCC's threading interface) implementations. In any case, the threading library is built with `gccCrossStageStatic` + `libcCross` / `lldClangNoCXX`, which works.

I declare this worth merging because I cross compiled Nix with it!

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @angerman @volth

### CODE REVIEW

**Comment by angerman** [comment]
This is pretty neat as it will likely allow to build `rocksdb`, however, we should note the following somewhere I believe:

> This project uses a number of undocumented NT system calls for efficiency reasons and may be broken in future Windows versions.
> The author gives no warranty for this project. Use it at your own risk.

I'm not sure what implications this might or might not incur wrt to building and distributing binaries for windows.

File: pkgs/os-specific/windows/mcfgthreads/default.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, fetchFromGitHub, autoreconfHook }:
+
+stdenv.mkDerivation {
+  pname = "mcfgthreads";
+  version = "git";
```
Comment (Mic92) [inline pkgs/os-specific/windows/mcfgthreads/default.nix]: We usually use date of commit here instead of `git`.

**Comment by Ericson2314** [comment]
@angerman I think it's probably fine for now; MS cares more about binary compat that anyone else, including these sorts of unstable interfaces. If it becomes a problem, it's all set up to be easy to switch to the pthreads emulation approach.

@Mic92 Thanks, will fix.

**Comment by angerman** [comment]
@Ericson2314 I'm running into lots of issues with this. I'm appreciative of this, however
- it doesn't compile on macOS, and such breaks macOS -> windows cross compilation for at least haskell.nix
- I can't get a linux -> windows rust compiler built with this.

Is there a global switch to turn this off?

**Pull request merged**